### PR TITLE
Reorganizing `stacked_borrows` in anticipation of a different model

### PR DESCRIPTION
**Pull request by Vanille-N**
These commits reorganize all the code inside the former `stacked_borrows` module and extract the part that really is specific to Stacked Borrows inside the `borrow_tracker/stacked_borrows` submodule.
Everything not specific to SB is put in `borrow_tracker/mod.rs`.
This is so that the future Tree Borrows model can be later added as a second submodule and reuse all the contents of `borrow_tracker/mod.rs`.

This reorganization is accompanied by renamings, mostly from "stacked borrows" to "borrow tracking".

### CODE REVIEW

**Comment by RalfJung** [comment]
Please don't include any merge commits in PRs. If there is a conflict, do a rebase instead.

File: src/borrow_tracker/mod.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        match self {
+            State::Stack(sb) => sb.borrow_mut().remove_unreachable_tags(tags),
+        }
+    }
+
+    pub fn expose_tag(&self, tag: BorTag) {
+        match self {
+            State::Stack(sb) => sb.borrow_mut().expose_tag(tag),
+        }
+    }
```
Comment (saethlin) [inline src/borrow_tracker/mod.rs:331]: The fact that these take `&self` is surprising. I think it'll make more sense now or later to hoist the `RefCell` out to `State` instead of sinking it into both the borrow tracking implementations.

File: src/borrow_tracker/mod.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+use std::fmt;
+use std::num::NonZeroU64;
+
+use log::trace;
+use smallvec::SmallVec;
+
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use stacked::diagnostics::RetagCause;
+
+use crate::*;
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:13]:
```suggestion

use stacked::diagnostics::RetagCause;
use crate::*;
```
The organization is
- std imports
- other dependency imports
- rustc imports
- local crate imports

with an empty line between each block.

File: src/borrow_tracker/mod.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    pub fn inner(&self) -> NonZeroU64 {
+        self.0
+    }
+
+    pub fn succ(self) -> Option<Self> {
+        self.try_map(|n| n.checked_add(1))
+    }
+
+    // The default to be used when SB is disabled
+    pub fn one() -> Self {
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:39]:
```suggestion
    pub fn default() -> Self {
```
No need to expose its value.

File: src/borrow_tracker/mod.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        Self::new(1).unwrap()
+    }
+
+    pub fn try_map<F>(self, f: F) -> Option<Self>
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:43]: What is this for? I think all you need is an increment function; if that is true, please just implement that directly.

File: src/borrow_tracker/mod.rs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+            }
+            trace!("New allocation {:?} has base tag {:?}", id, tag);
+            self.base_ptr_tags.try_insert(id, tag).unwrap();
+            tag
+        })
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Method {
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:225]: This is missing a doc comment.

File: src/borrow_tracker/mod.rs [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+                )))),
+            Method::Off => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum State {
+    Stack(stacked::AllocExtra),
+}
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:270]: This is missing a doc comment. it is also a confusing name... we have GlobalState and State?

What about calling this AllocState or so? Should then probably rename `stacked::AllocExtra` to `stacked::AllocState` for consistency.

File: src/borrow_tracker/mod.rs [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+impl State {
+    pub fn assume_sb(&self) -> &stacked::AllocExtra {
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:273]: `assert_sb` I think would be more in line with how we usually call these -- I assume this will panic if the state is not sb.

File: src/borrow_tracker/mod.rs [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+                    machine,
+                )))),
+            Method::Off => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum State {
+    Stack(stacked::AllocExtra),
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:269]: Why does this not have a `None` variant corresponding to `Method::Off`?

File: src/borrow_tracker/stacked/mod.rs [hunk 9]
```diff
@@ -4,80 +4,34 @@
...
+    stacked::diagnostics::{AllocHistory, DiagnosticCx, DiagnosticCxBuilder, TagHistory},
+    AccessKind, GlobalState, GlobalStateInner, ProtectorKind, RetagCause, RetagFields,
+};
+use rustc_middle::{
+    mir::{Mutability, RetagKind},
+    ty::{
+        self,
+        layout::{HasParamEnv, LayoutOf},
+    },
+};
```
Comment (RalfJung) [inline src/borrow_tracker/stacked/mod.rs:23]: Please keep the import structure, and move this up to the other rustc imports.

File: src/borrow_tracker/mod.rs (hunk 1 above)
Comment (RalfJung) [reply src/borrow_tracker/mod.rs:328]: We'll want the `RefCell` to be inside the `Option` though, so if we add a `None` variant to `State` then we wouldn't want a `RefCell` around that.

Comment (RalfJung) [reply src/borrow_tracker/mod.rs:328]: Actually this could just take `&mut self` I think, the only caller has a mutable reference available.

But I agree that since this file here is doing the `borrow_mut` calls, this file should also be adding the `RefCell`. We are already assuming that `stacked::AllocExtra` is a `RefCell`. So might as well define this more like
```rust
pub enum AllocState {
    Stack(RefCell<stacked::AllocState>),
    None,
}
```

File: src/borrow_tracker/stacked/mod.rs [hunk 10]
```diff
@@ -929,14 +754,16 @@ trait EvalContextPrivExt<'mir: 'ecx, 'tcx: 'mir, 'ecx>: crate::MiriInterpCxExt<'
...
             .as_mut()
             .expect("we should have Stacked Borrows data")
-            .get_mut();
+            .assume_sb()
+            .borrow_mut();
         let item = Item::new(new_tag, perm, protect.is_some());
         let range = alloc_range(base_offset, size);
-        let global = machine.stacked_borrows.as_ref().unwrap().borrow();
+        let global = machine.borrow_tracker_data.as_ref().unwrap().borrow();
+        // FIXME: can't share this with the current_span inside log_creation
```
Comment (RalfJung) [inline src/borrow_tracker/stacked/mod.rs:766]: You are reverting some recent changes in master here. Please be careful when rebasing!

File: src/borrow_tracker/stacked/mod.rs [hunk 11]
```diff
@@ -961,7 +788,8 @@ trait EvalContextPrivExt<'mir: 'ecx, 'tcx: 'mir, 'ecx>: crate::MiriInterpCxExt<'
 
     /// Retags an indidual pointer, returning the retagged version.
     /// `mutbl` can be `None` to make this a raw pointer.
-    fn retag_reference(
+    // FIXME: is this comment outdated ?
```
Comment (RalfJung) [inline src/borrow_tracker/stacked/mod.rs:791]:
```suggestion
```

File: src/borrow_tracker/stacked/mod.rs [hunk 12]
```diff
@@ -961,7 +788,8 @@ trait EvalContextPrivExt<'mir: 'ecx, 'tcx: 'mir, 'ecx>: crate::MiriInterpCxExt<'
 
     /// Retags an indidual pointer, returning the retagged version.
     /// `mutbl` can be `None` to make this a raw pointer.
```
Comment (RalfJung) [inline src/borrow_tracker/stacked/mod.rs:790]:
```suggestion
    /// Retags an indidual pointer, returning the retagged version.
    /// `kind` indicates what kind of reference is being created.
```

File: src/borrow_tracker/stacked/mod.rs [hunk 13]
```diff
@@ -1181,7 +1014,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 // uncovers a non-supported `extern static`.
                 let alloc_extra = this.get_alloc_extra(alloc_id)?;
                 trace!("Stacked Borrows tag {tag:?} exposed in {alloc_id:?}");
-                alloc_extra.stacked_borrows.as_ref().unwrap().borrow_mut().exposed_tags.insert(tag);
+                alloc_extra.borrow_tracker.as_ref().unwrap().expose_tag(tag);
```
Comment (RalfJung) [inline src/borrow_tracker/stacked/mod.rs:1017]: Why does Stacked Borrows call back into the generic borrow tracking machinery here? This should call `assume_sb`.

File: src/borrow_tracker/stacked/mod.rs [hunk 14]
```diff
@@ -775,12 +598,13 @@ trait EvalContextPrivExt<'mir: 'ecx, 'tcx: 'mir, 'ecx>: crate::MiriInterpCxExt<'
                         orig_tag,
                         alloc_range(base_offset, size),
                     );
-                    let mut dcx = dcx.build(&mut stacked_borrows.history, base_offset);
+                    let mut stacked_borrows = sb.borrow_mut();
+                    let mut dcx = dcx.build(stacked_borrows.history_mut(), base_offset);
```
Comment (RalfJung) [inline src/borrow_tracker/stacked/mod.rs:602]: Why did you add a new method for this?

File: src/machine.rs [hunk 15]
```diff
@@ -351,7 +351,8 @@ pub struct MiriMachine<'mir, 'tcx> {
     pub tcx: TyCtxt<'tcx>,
 
     /// Stacked Borrows global data.
-    pub stacked_borrows: Option<stacked_borrows::GlobalState>,
+    pub borrow_tracker_method: BorrowTrackerMethod,
+    pub borrow_tracker_data: Option<borrow_tracker::GlobalState>,
```
Comment (RalfJung) [inline src/machine.rs:355]: `borrow_tracker_method` should not be needed, since it is entirely determined by `borrow_tracker_data`. Then you can rename that field to `borrow_tracker`.

File: src/machine.rs [hunk 16]
```diff
@@ -477,17 +478,13 @@ impl<'mir, 'tcx> MiriMachine<'mir, 'tcx> {
...
         let rng = StdRng::seed_from_u64(config.seed.unwrap_or(0));
-        let stacked_borrows = config.stacked_borrows.then(|| {
-            RefCell::new(stacked_borrows::GlobalStateInner::new(
-                config.tracked_pointer_tags.clone(),
-                config.tracked_call_ids.clone(),
-                config.retag_fields,
-            ))
-        });
+        let borrow_tracker_method = config.borrow_tracker;
+        let borrow_tracker_data = borrow_tracker_method.instanciate_global_state(config);
```
Comment (RalfJung) [inline src/machine.rs:482]:
```suggestion
        let borrow_tracker = config.borrow_tracker.instantiate_global_state(config);
```

File: src/machine.rs [hunk 17]
```diff
@@ -900,9 +898,13 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
         }
 
         let alloc = alloc.into_owned();
-        let stacks = ecx.machine.stacked_borrows.as_ref().map(|stacked_borrows| {
-            Stacks::new_allocation(id, alloc.size(), stacked_borrows, kind, &ecx.machine)
-        });
+        let borrow_tracker = ecx.machine.borrow_tracker_method.new_allocation(
```
Comment (RalfJung) [inline src/machine.rs:901]: This can probably be `ecx.machine.borrow_tracker.method().new_allocation`.

File: src/machine.rs [hunk 18]
```diff
@@ -1068,7 +1059,10 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
         kind: mir::RetagKind,
         place: &PlaceTy<'tcx, Provenance>,
     ) -> InterpResult<'tcx> {
-        if ecx.machine.stacked_borrows.is_some() { ecx.retag(kind, place) } else { Ok(()) }
+        match ecx.machine.borrow_tracker_method {
+            BorrowTrackerMethod::StackedBorrows => ecx.sb_retag(kind, place),
+            BorrowTrackerMethod::Off => Ok(()),
+        }
```
Comment (RalfJung) [inline src/machine.rs:1065]: Please move this `match` into `borrow_tracker/mod.rs`.
The machine file should never match on the `BorrowTrackerMethod`.

File: src/machine.rs [hunk 19]
```diff
@@ -1091,10 +1085,10 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
             None
         };
 
-        let stacked_borrows = ecx.machine.stacked_borrows.as_ref();
+        let borrow_tracker = ecx.machine.borrow_tracker_data.as_ref();
 
         let extra = FrameData {
-            stacked_borrows: stacked_borrows.map(|sb| sb.borrow_mut().new_frame(&ecx.machine)),
+            borrow_tracker: borrow_tracker.map(|sb| sb.borrow_mut().new_frame(&ecx.machine)),
```
Comment (RalfJung) [inline src/machine.rs:1091]:
```suggestion
            borrow_tracker: borrow_tracker.map(|bt| bt.borrow_mut().new_frame(&ecx.machine)),
```

File: src/machine.rs [hunk 20]
```diff
@@ -1154,7 +1148,11 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
             ecx.active_thread_mut().set_top_user_relevant_frame(stack_len - 1);
         }
 
-        if ecx.machine.stacked_borrows.is_some() { ecx.retag_return_place() } else { Ok(()) }
+        match ecx.machine.borrow_tracker_method {
```
Comment (RalfJung) [inline src/machine.rs:1151]: Another `match` to move into the borrow tracker.

File: src/machine.rs [hunk 21]
```diff
@@ -1154,7 +1148,11 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
...
+            BorrowTrackerMethod::StackedBorrows => ecx.sb_retag_return_place(),
+            // FIXME: don't forget to also activate the args after entry
```
Comment (RalfJung) [inline src/machine.rs:1153]: I don't quite understand this FIXME?

File: src/borrow_tracker/mod.rs (hunk 6 above)
Comment (Vanille-N) [reply src/borrow_tracker/mod.rs:270]: Or I could call it `AllocExtra` as it dispatches into each implementation's `AllocExtra` ? Or would that be too confusing ?

File: src/borrow_tracker/stacked/mod.rs (hunk 10 above)
Comment (Vanille-N) [reply src/borrow_tracker/stacked/mod.rs]: `get_mut` doesn't work with `assert_sb` though. Or is that not the issue ?

File: src/borrow_tracker/mod.rs (hunk 1 above)
Comment (Vanille-N) [reply src/borrow_tracker/mod.rs]: Ok I'll do that

File: src/borrow_tracker/mod.rs (hunk 6 above)
Comment (RalfJung) [reply src/borrow_tracker/mod.rs]: We also call it `GlobalState`, not `GlobalExtra`. So I think we should rename the implementation's thing to `AllocState`, too.

File: src/borrow_tracker/stacked/mod.rs (hunk 10 above)
Comment (RalfJung) [reply src/borrow_tracker/stacked/mod.rs]: Ah, then you need `assert_sb_mut`.

**Comment by RalfJung** [comment]
@rustbot author

(call `rustbot ready` when this is ready for another round of review)

**Review by RalfJung — COMMENTED** [review]
There are also still some unresolved comments from the previous round. (Github might hide them for you, be sure to click "load more" until that button disappears.)

File: src/borrow_tracker/stacked/mod.rs [hunk 22]
```diff
@@ -1181,7 +989,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 // uncovers a non-supported `extern static`.
                 let alloc_extra = this.get_alloc_extra(alloc_id)?;
                 trace!("Stacked Borrows tag {tag:?} exposed in {alloc_id:?}");
-                alloc_extra.stacked_borrows.as_ref().unwrap().borrow_mut().exposed_tags.insert(tag);
+                alloc_extra.borrow_tracker.expose_tag(tag);
```
Comment (RalfJung) [inline src/borrow_tracker/stacked/mod.rs:992]: As I already said last round:
Why does Stacked Borrows call back into the generic borrow tracking machinery here? This should call assume_sb. Then you can probably remove the new expose_tag method on the per-alloc state.

File: src/intptrcast.rs [hunk 23]
```diff
@@ -105,15 +105,16 @@ impl<'mir, 'tcx> GlobalStateInner {
...
         let global_state = ecx.machine.intptrcast.get_mut();
         // In strict mode, we don't need this, so we can save some cycles by not tracking it.
         if global_state.provenance_mode != ProvenanceMode::Strict {
             trace!("Exposing allocation id {alloc_id:?}");
             global_state.exposed.insert(alloc_id);
-            if ecx.machine.stacked_borrows.is_some() {
-                ecx.expose_tag(alloc_id, sb)?;
+            match ecx.machine.borrow_tracker_method {
+                BorrowTrackerMethod::StackedBorrows => ecx.sb_expose_tag(alloc_id, tag)?,
+                BorrowTrackerMethod::Off => (),
```
Comment (RalfJung) [inline src/intptrcast.rs:117]: As I said last time, there shouldn't be any `match` on this outside the borrow_tracker module.

File: src/machine.rs (hunk 15 above)
Comment (RalfJung) [reply src/machine.rs:356]: Oh I see, the global borrow_tracker_data is the same no matter the borrow tracking module.

But still, we should only have one field for the borrow tracker here. `new_allocation` should become a function on `borrow_tracker::GlobalState` which then has to internally remember whatever is needed.

File: src/intptrcast.rs (hunk 23 above)
Comment (Vanille-N) [reply src/intptrcast.rs:117]: Oops sorry I missed that one

File: src/borrow_tracker/mod.rs [hunk 24]
```diff
@@ -0,0 +1,370 @@
...
+impl VisitTags for FrameExtra {
+    fn visit_tags(&self, _visit: &mut dyn FnMut(BorTag)) {
+        // `protected_tags` are fine to GC.
+    }
+}
+
+/// Extra global state, available to the memory access hooks.
+#[derive(Debug)]
+pub struct GlobalStateInner {
+    /// Borrow tracker method currently in use (except BorrowTrackerMethod::Off)
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:84]:
```suggestion
    /// Borrow tracker method currently in use.
```

File: src/borrow_tracker/mod.rs [hunk 25]
```diff
@@ -274,20 +274,20 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         &mut self,
         kind: RetagKind,
         place: &PlaceTy<'tcx, Provenance>,
-        borrow_tracker_method: BorrowTrackerMethod,
     ) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
-        match borrow_tracker_method {
+        let method = this.machine.borrow_tracker.as_ref().unwrap().borrow().method();
```
Comment (RalfJung) [inline src/borrow_tracker/mod.rs:275]: Why do you call `method()` here? You should have access to the private field, it's in the same module. (Same below a few more times. Is `fn method` even needed?)

File: src/borrow_tracker/stacked_borrows/mod.rs [hunk 26]
```diff
@@ -693,12 +494,16 @@ impl Stacks {
...
         let dcx = DiagnosticCxBuilder::dealloc(machine, tag);
-        let state = machine.stacked_borrows.as_ref().unwrap().borrow();
+        let state = machine.borrow_tracker.as_ref().unwrap().borrow();
         self.for_each(range, dcx, |stack, dcx, exposed_tags| {
             stack.dealloc(tag, &state, dcx, exposed_tags)
         })?;
         Ok(())
     }
+
+    pub fn expose_tag(&mut self, tag: BorTag) {
```
Comment (RalfJung) [inline src/borrow_tracker/stacked_borrows/mod.rs:504]:
```suggestion
    fn expose_tag(&mut self, tag: BorTag) {
```
This is only called in this file, right?

**Comment by RalfJung** [comment]
Looks good. :) Just a few final nits. Then please squash this into one (or a few) commits and we are good to go!

**Comment by RalfJung** [comment]
(force-pushes do not always generate GH notifications, so make sure to always post a note in the PR if you want another review)

**Comment by RalfJung** [comment]
Thanks!

@bors r+
